### PR TITLE
feat(ingress)!: Support for multiple ingress routes and TCP connections

### PR DESCRIPTION
**Pull request by gburnard**
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Firmware version:
* Hardware:
* Toolchain:
* SDK:

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: packages/c6o-system/src/mixins/api/createAppPost.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@ export const createAppPostMixin = (base: baseProvisionerType) => class extends b
 
     async createAppPost(app: AppDocument) {
 
-        if (app.spec.routes?.simple) {
+        if (app.spec.routes?.length !== 0) {
```
Comment (nsainaney) [inline packages/c6o-system/src/mixins/api/createAppPost.ts:8]:
```suggestion
        if (app.spec.routes?.length) {
```

File: packages/common/src/app.ts [hunk 2]
```diff
@@ -18,14 +18,15 @@ export interface LaunchType {
...
-        http?: {
-            service: string
-            port?: number
-            prefix?: string
-            rewrite?: string
-        }
-    }
+    name: string,
+    type: 'tcp'|'http',
+    service: string,
```
Comment (nsainaney) [inline packages/common/src/app.ts:23]:
```suggestion
    serviceName: string,
```

File: packages/c6o-system/src/index.ts [hunk 3]
```diff
@@ -6,7 +6,7 @@ import {
     updateApplyMixin,
     updateSystemMixin,
     createValidateMixin,
-    createAppPostMixin,
+    routingMixin,
```
Comment (nsainaney) [inline packages/c6o-system/src/index.ts:9]:
```suggestion
    postAppMixin,
```

File: packages/c6o-system/src/index.ts [hunk 4]
```diff
@@ -18,7 +18,7 @@ import {
 export type baseProvisionerType = new (...a) => Provisioner & ProvisionerBase
 
 export class Provisioner extends mix(ProvisionerBase).with(
-    createAppPostMixin,
+    routingMixin,
```
Comment (nsainaney) [inline packages/c6o-system/src/index.ts:21]:
```suggestion
    postAppMixin,
```

File: packages/c6o-system/src/mixins/api/createAppPost.ts [hunk 5]
```diff
@@ -1,15 +1,44 @@
 import { AppDocument } from '@provisioner/common'
 import { baseProvisionerType } from '../../'
 
-export const createAppPostMixin = (base: baseProvisionerType) => class extends base {
+export const routingMixin = (base: baseProvisionerType) => class extends base {
```
Comment (nsainaney) [inline packages/c6o-system/src/mixins/api/createAppPost.ts:4]:
```suggestion
export const postAppMixin = (base: baseProvisionerType) => class extends base {
```

File: packages/common/src/app.ts [hunk 6]
```diff
@@ -41,7 +46,7 @@ export interface AppDocument extends KubeDocument {
         }
         provisioner?: any | 'ignore'
         services?: ServicesType
-        routes?: RoutesType
+        routes?: Array<RoutesType>
```
Comment (nsainaney) [inline packages/common/src/app.ts:49]:
```suggestion
        routes?: RoutesType[]
```

File: packages/c6o-system/src/mixins/api/createAppPost.ts [hunk 7]
```diff
@@ -1,15 +1,44 @@
...
             const istioProvisioner = await this.manager.getAppProvisioner('istio', 'istio-system')
             await istioProvisioner.createVirtualService(app, 'c6o-system/' + this.SYSTEM_GATEWAY_NAME)
             this.manager.status?.pop()
         }
     }
-}
+
+    async postRemoveApp(app: AppDocument) {
+        if (app.spec.routes?.length !== 0) {
+            this.manager.status?.push(`Tearing down app ${app.metadata.namespace} routes`)
```
Comment (nsainaney) [inline packages/c6o-system/src/mixins/api/createAppPost.ts:17]:
```suggestion
            this.manager.status?.push(`Removing App ${app.metadata.namespace} routes`)
```

File: packages/c6o-system/src/mixins/api/createAppPost.ts [hunk 8]
```diff
@@ -1,15 +1,44 @@
...
 
-export const createAppPostMixin = (base: baseProvisionerType) => class extends base {
+export const routingMixin = (base: baseProvisionerType) => class extends base {
 
-    async createAppPost(app: AppDocument) {
-
-        if (app.spec.routes?.simple) {
+    async postCreateApp(app: AppDocument) {
+        if (app.spec.routes?.length !== 0) {
             this.manager.status?.push(`Setting up app ${app.metadata.namespace} routes`)
```
Comment (nsainaney) [inline packages/c6o-system/src/mixins/api/createAppPost.ts:8]:
```suggestion
            this.manager.status?.push(`Creating App ${app.metadata.namespace} routes`)
```

File: packages/c6o-system/src/mixins/api/createAppPost.ts [hunk 9]
```diff
@@ -1,15 +1,44 @@
...
+            this.manager.status?.push(`Tearing down app ${app.metadata.namespace} routes`)
+            const istioProvisioner = await this.manager.getAppProvisioner('istio', 'istio-system')
+            await istioProvisioner.removeVirtualService(app)
+            this.manager.status?.pop()
+        }
+    }
+
+    async postUpdateApp(app: AppDocument) {
+        if (app.spec.routes?.length !== 0) {
+            this.manager.status?.push(`Updating app ${app.metadata.namespace} routes`)
```
Comment (nsainaney) [inline packages/c6o-system/src/mixins/api/createAppPost.ts:26]:
```suggestion
            this.manager.status?.push(`Updating App ${app.metadata.namespace} routes`)
```

**Review by nsainaney — CHANGES_REQUESTED** [review]

File: packages/c6o-system/src/mixins/api/createAppPost.ts (hunk 1 above)
Comment (gburnard) [reply packages/c6o-system/src/mixins/api/createAppPost.ts:7]: done

File: packages/common/src/app.ts (hunk 2 above)
Comment (gburnard) [reply packages/common/src/app.ts]: done

File: packages/c6o-system/src/index.ts (hunk 3 above)
Comment (gburnard) [reply packages/c6o-system/src/index.ts:9]: done

File: packages/c6o-system/src/index.ts (hunk 4 above)
Comment (gburnard) [reply packages/c6o-system/src/index.ts:21]: done

File: packages/c6o-system/src/mixins/api/createAppPost.ts (hunk 5 above)
Comment (gburnard) [reply packages/c6o-system/src/mixins/api/createAppPost.ts:4]: done

File: packages/common/src/app.ts (hunk 6 above)
Comment (gburnard) [reply packages/common/src/app.ts:49]: What about marina.menuItems @nsainaney

File: packages/c6o-system/src/mixins/api/createAppPost.ts (hunk 8 above)
Comment (gburnard) [reply packages/c6o-system/src/mixins/api/createAppPost.ts:8]: done

File: packages/c6o-system/src/mixins/api/createAppPost.ts (hunk 9 above)
Comment (gburnard) [reply packages/c6o-system/src/mixins/api/createAppPost.ts:26]: done

**Review by nsainaney — APPROVED** [review]

**Pull request merged**